### ISSUE TITLE
Добавить расчет когнитивной и цикломатической сложностей

### ISSUE THREAD

**Issue by ilya2184** [1C-Company/v8-code-style#1117]
### Описание проблемы

Было ба не плохо если бы в проверках "по стандартам" (я понимаю что стандарта нет такого) были два пункта:
Предельное значение когнитивной сложности
Предельное значение цикломатической сложности

При превышении указанных значений добавлялась незначительная ошибка конфигурации.

### Описание решения проблемы

Предлагаю сделать так как это реализовано в BSL Language Server:
[Когнитивная сложность (CognitiveComplexity)](https://1c-syntax.github.io/bsl-language-server/diagnostics/CognitiveComplexity/)
[Цикломатическая сложность (CyclomaticComplexity)](https://1c-syntax.github.io/bsl-language-server/diagnostics/CyclomaticComplexity/)

**Comment by marmyshev** [comment]
Исходные ссылки на [Когнитивную сложность](https://www.sonarsource.com/docs/CognitiveComplexity.pdf) и [Википедия](https://en.wikipedia.org/wiki/Cognitive_complexity)

**Comment by marmyshev** [comment]
Исходные ссылки на циклометрическую сложность  [Википедия](https://en.wikipedia.org/wiki/Cyclomatic_complexity) [Википедия RU](https://ru.wikipedia.org/wiki/Цикломатическая_сложность)

**Comment by RedMammoth** [comment]
Что-то у меня разошлись значения когнитивной сложности в примере2:
bls-ls говорит 33, а если подсчитать по комментам, вроде как 40.
![image](https://user-images.githubusercontent.com/64589646/224515237-7e153d25-8df1-4736-8bc6-963d7f84f781.png)

Поэтому решил исходить из спецификации (Приложение B) источника (https://www.sonarsource.com/docs/CognitiveComplexity.pdf)

Буду придерживать следующего алгоритма, поправьте, если допустил ошибки:

Конструкции увеличивающие сложность:
- Если, ИначеЕсли, Иначе
- тренарный оператор ?(,,)
- циклы
- ветвь исключения в попытке
- переход к метке
- логические операнды И, ИЛИ
- рекурсивный вызов

Конструкции получающие дополнительный штраф за уровень вложенности:
- если
- тренарный оператор
- циклы
- ветвь исключения в попытке

Конструкции увеличивающие уровень вложенности:
- если, ИначеЕсли, Иначе
- тренарный оператор
- циклы
- ветвь исключения в попытке
- вложенный вызов метода

**Comment by DmitryShvaika** [comment]
А Edt умеет подсчитывать когнитивную и цикломатическую сложность? Не подскажите где эти настройки? И можно ли выдавать ошибку при превышении заданного порога?

**Comment by RedMammoth** [comment]
> А Edt умеет подсчитывать когнитивную и цикломатическую сложность? Не подскажите где эти настройки? И можно ли выдавать ошибку при превышении заданного порога?

Нет, ещё не умеет, как раз пробую научить. На скрине VCS с плагином

**Comment by RedMammoth** [comment]
Взял в работу

**Comment by RedMammoth** [comment]
Приведенный алгоритм цикломатической сложности тоже не нравится https://1c-syntax.github.io/bsl-language-server/diagnostics/CyclomaticComplexity/

Мне кажется мы не должны увеличивать счетчик для блока Иначе, также мне кажется цикломатическая сложность не возрастает при переходах goto. Зато, скорей всего необходимо учесть директивы компиляции

**Comment by marmyshev** [comment]
Блок "Иначе" - это альтернативный (ещё один) вариант условий. Т.е. человек необходимо понять все условия срабатывания в "Если" или в "Иначе если" - потом понять весь скоуп всех возможных вариантов, вычесть из них те что могли отработать выше - и понять в каких же случаях реально будет срабатывание кода в "Иначе".

Это однозначное усложнение.

**Comment by marmyshev** [comment]
С метками goto всё ещё хуже: человек перестаёт понимать алгоритм линейно пока читает код, ему нужно перепрыгивать в другое место чтобы понимать и переносить (в своей голове) туда весь текущий контекст код, при чем он не может опираться на тот код который видит до метки, так как этого кода в данный момент (при переходе) не существует как бы.

То есть уровень загрузки/напряга мозга сильно повышается на метках.

**Comment by RedMammoth** [comment]
> Блок "Иначе" - это альтернативный (ещё один) вариант условий. Т.е. человек необходимо понять все условия срабатывания в "Если" или в "Иначе если" - потом понять весь скоуп всех возможных вариантов, вычесть из них те что могли отработать выше - и понять в каких же случаях реально будет срабатывание кода в "Иначе".
> 
> Это однозначное усложнение.

У нас Если добавляет ещё одну линию, вне зависимости от наличия Иначе, просто в сокращённом случае блок если пустой и опускается. Поэтому наличие если третью линию не добавляет. А вот ветки ИначеЕсли каждая добавляет по линии

**Comment by RedMammoth** [comment]
> С метками goto всё ещё хуже: человек перестаёт понимать алгоритм линейно пока читает код, ему нужно перепрыгивать в другое место чтобы понимать и переносить (в своей голове) туда весь текущий контекст код, при чем он не может опираться на тот код который видит до метки, так как этого кода в данный момент (при переходе) не существует как бы.
> 
> То есть уровень загрузки/напряга мозга сильно повышается на метках.

То что мозг напряжется согласен, но изменится ли у нас количество линий исполнения?
Если правильно понимаю смысл цикломатической проверки, она должна сказать сколько минимум тестов нужно написать на код. Для это и подсчитываются линии исполнения. Можно ли подсчитать, сколько тестов добавит наличие перехода к метке?
